In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [3]:
dataset = pd.read_csv("F:\\LetsUpgrade\\Notes\\DAY 7 Notes + Ass\\general_data.csv")

In [5]:
dataset.head()

Age Attrition     BusinessTravel              Department  DistanceFromHome  \
0   51        No      Travel_Rarely                   Sales                 6   
1   31       Yes  Travel_Frequently  Research & Development                10   
2   32        No  Travel_Frequently  Research & Development                17   
3   38        No         Non-Travel  Research & Development                 2   
4   32        No      Travel_Rarely  Research & Development                10   

   Education EducationField  EmployeeCount  EmployeeID  Gender  ...  \
0          2  Life Sciences              1           1  Female  ...   
1          1  Life Sciences              1           2  Female  ...   
2          4          Other              1           3    Male  ...   
3          5  Life Sciences              1           4    Male  ...   
4          1        Medical              1           5    Male  ...   

   NumCompaniesWorked Over18 PercentSalaryHike  StandardHours  \
0                 1.0      Y                11              8   
1                 0.0      Y                23              8   
2                 1.0      Y                15              8   
3                 3.0      Y                11              8   
4                 4.0      Y                12              8   

   StockOptionLevel TotalWorkingYears  TrainingTimesLastYear  YearsAtCompany  \
0                 0               1.0                      6               1   
1                 1               6.0                      3               5   
2                 3               5.0                      2               5   
3                 3              13.0                      5               8   
4                 2               9.0                      2               6   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        0                     0  
1                        1                     4  
2                        0                     3  
3                        7                     5  
4                        0                     4  

[5 rows x 24 columns]

In [7]:
dummy=pd.get_dummies(dataset['Attrition'])
dummy

No  Yes
0      1    0
1      0    1
2      1    0
3      1    0
4      1    0
...   ..  ...
4405   1    0
4406   1    0
4407   1    0
4408   1    0
4409   1    0

[4410 rows x 2 columns]

In [8]:
dataset2=pd.concat((dataset,dummy),axis=1)
dataset2

Age Attrition     BusinessTravel              Department  \
0      51        No      Travel_Rarely                   Sales   
1      31       Yes  Travel_Frequently  Research & Development   
2      32        No  Travel_Frequently  Research & Development   
3      38        No         Non-Travel  Research & Development   
4      32        No      Travel_Rarely  Research & Development   
...   ...       ...                ...                     ...   
4405   42        No      Travel_Rarely  Research & Development   
4406   29        No      Travel_Rarely  Research & Development   
4407   25        No      Travel_Rarely  Research & Development   
4408   42        No      Travel_Rarely                   Sales   
4409   40        No      Travel_Rarely  Research & Development   

      DistanceFromHome  Education EducationField  EmployeeCount  EmployeeID  \
0                    6          2  Life Sciences              1           1   
1                   10          1  Life Sciences              1           2   
2                   17          4          Other              1           3   
3                    2          5  Life Sciences              1           4   
4                   10          1        Medical              1           5   
...                ...        ...            ...            ...         ...   
4405                 5          4        Medical              1        4406   
4406                 2          4        Medical              1        4407   
4407                25          2  Life Sciences              1        4408   
4408                18          2        Medical              1        4409   
4409                28          3        Medical              1        4410   

      Gender  ...  PercentSalaryHike StandardHours StockOptionLevel  \
0     Female  ...                 11             8                0   
1     Female  ...                 23             8                1   
2       Male  ...                 15             8                3   
3       Male  ...                 11             8                3   
4       Male  ...                 12             8                2   
...      ...  ...                ...           ...              ...   
4405  Female  ...                 17             8                1   
4406    Male  ...                 15             8                0   
4407    Male  ...                 20             8                0   
4408    Male  ...                 14             8                1   
4409    Male  ...                 12             8                0   

      TotalWorkingYears  TrainingTimesLastYear YearsAtCompany  \
0                   1.0                      6              1   
1                   6.0                      3              5   
2                   5.0                      2              5   
3                  13.0                      5              8   
4                   9.0                      2              6   
...                 ...                    ...            ...   
4405               10.0                      5              3   
4406               10.0                      2              3   
4407                5.0                      4              4   
4408               10.0                      2              9   
4409                NaN                      6             21   

      YearsSinceLastPromotion  YearsWithCurrManager  No  Yes  
0                           0                     0   1    0  
1                           1                     4   0    1  
2                           0                     3   1    0  
3                           7                     5   1    0  
4                           0                     4   1    0  
...                       ...                   ...  ..  ...  
4405                        0                     2   1    0  
4406                        0                     2   1    0  
4407                        1                     2   1    0  


In [9]:
dataset2=dataset2.drop(['Attrition'],axis=1)
dataset2

Age     BusinessTravel              Department  DistanceFromHome  \
0      51      Travel_Rarely                   Sales                 6   
1      31  Travel_Frequently  Research & Development                10   
2      32  Travel_Frequently  Research & Development                17   
3      38         Non-Travel  Research & Development                 2   
4      32      Travel_Rarely  Research & Development                10   
...   ...                ...                     ...               ...   
4405   42      Travel_Rarely  Research & Development                 5   
4406   29      Travel_Rarely  Research & Development                 2   
4407   25      Travel_Rarely  Research & Development                25   
4408   42      Travel_Rarely                   Sales                18   
4409   40      Travel_Rarely  Research & Development                28   

      Education EducationField  EmployeeCount  EmployeeID  Gender  JobLevel  \
0             2  Life Sciences              1           1  Female         1   
1             1  Life Sciences              1           2  Female         1   
2             4          Other              1           3    Male         4   
3             5  Life Sciences              1           4    Male         3   
4             1        Medical              1           5    Male         1   
...         ...            ...            ...         ...     ...       ...   
4405          4        Medical              1        4406  Female         1   
4406          4        Medical              1        4407    Male         1   
4407          2  Life Sciences              1        4408    Male         2   
4408          2        Medical              1        4409    Male         1   
4409          3        Medical              1        4410    Male         2   

      ... PercentSalaryHike StandardHours  StockOptionLevel  \
0     ...                11             8                 0   
1     ...                23             8                 1   
2     ...                15             8                 3   
3     ...                11             8                 3   
4     ...                12             8                 2   
...   ...               ...           ...               ...   
4405  ...                17             8                 1   
4406  ...                15             8                 0   
4407  ...                20             8                 0   
4408  ...                14             8                 1   
4409  ...                12             8                 0   

      TotalWorkingYears TrainingTimesLastYear  YearsAtCompany  \
0                   1.0                     6               1   
1                   6.0                     3               5   
2                   5.0                     2               5   
3                  13.0                     5               8   
4                   9.0                     2               6   
...                 ...                   ...             ...   
4405               10.0                     5               3   
4406               10.0                     2               3   
4407                5.0                     4               4   
4408               10.0                     2               9   
4409                NaN                     6              21   

      YearsSinceLastPromotion  YearsWithCurrManager  No  Yes  
0                           0                     0   1    0  
1                           1                     4   0    1  
2                           0                     3   1    0  
3                           7                     5   1    0  
4                           0                     4   1    0  
...                       ...                   ...  ..  ...  
4405                        0                     2   1    0  
4406                        0                     2   1    0  
4407                        1                     2   1    0  


In [10]:
dataset2 = dataset2.drop(["No"],axis=1)
dataset2

Age     BusinessTravel              Department  DistanceFromHome  \
0      51      Travel_Rarely                   Sales                 6   
1      31  Travel_Frequently  Research & Development                10   
2      32  Travel_Frequently  Research & Development                17   
3      38         Non-Travel  Research & Development                 2   
4      32      Travel_Rarely  Research & Development                10   
...   ...                ...                     ...               ...   
4405   42      Travel_Rarely  Research & Development                 5   
4406   29      Travel_Rarely  Research & Development                 2   
4407   25      Travel_Rarely  Research & Development                25   
4408   42      Travel_Rarely                   Sales                18   
4409   40      Travel_Rarely  Research & Development                28   

      Education EducationField  EmployeeCount  EmployeeID  Gender  JobLevel  \
0             2  Life Sciences              1           1  Female         1   
1             1  Life Sciences              1           2  Female         1   
2             4          Other              1           3    Male         4   
3             5  Life Sciences              1           4    Male         3   
4             1        Medical              1           5    Male         1   
...         ...            ...            ...         ...     ...       ...   
4405          4        Medical              1        4406  Female         1   
4406          4        Medical              1        4407    Male         1   
4407          2  Life Sciences              1        4408    Male         2   
4408          2        Medical              1        4409    Male         1   
4409          3        Medical              1        4410    Male         2   

      ... Over18 PercentSalaryHike  StandardHours  StockOptionLevel  \
0     ...      Y                11              8                 0   
1     ...      Y                23              8                 1   
2     ...      Y                15              8                 3   
3     ...      Y                11              8                 3   
4     ...      Y                12              8                 2   
...   ...    ...               ...            ...               ...   
4405  ...      Y                17              8                 1   
4406  ...      Y                15              8                 0   
4407  ...      Y                20              8                 0   
4408  ...      Y                14              8                 1   
4409  ...      Y                12              8                 0   

     TotalWorkingYears  TrainingTimesLastYear  YearsAtCompany  \
0                  1.0                      6               1   
1                  6.0                      3               5   
2                  5.0                      2               5   
3                 13.0                      5               8   
4                  9.0                      2               6   
...                ...                    ...             ...   
4405              10.0                      5               3   
4406              10.0                      2               3   
4407               5.0                      4               4   
4408              10.0                      2               9   
4409               NaN                      6              21   

      YearsSinceLastPromotion  YearsWithCurrManager  Yes  
0                           0                     0    0  
1                           1                     4    1  
2                           0                     3    0  
3                           7                     5    0  
4                           0                     4    0  
...                       ...                   ...  ...  
4405                        0                     2    0  
4406                        0                     2    0  
440

In [12]:
dataset2=dataset2.rename(columns={"Yes":"Attrition"})
dataset2.head()

Age     BusinessTravel              Department  DistanceFromHome  \
0   51      Travel_Rarely                   Sales                 6   
1   31  Travel_Frequently  Research & Development                10   
2   32  Travel_Frequently  Research & Development                17   
3   38         Non-Travel  Research & Development                 2   
4   32      Travel_Rarely  Research & Development                10   

   Education EducationField  EmployeeCount  EmployeeID  Gender  JobLevel  ...  \
0          2  Life Sciences              1           1  Female         1  ...   
1          1  Life Sciences              1           2  Female         1  ...   
2          4          Other              1           3    Male         4  ...   
3          5  Life Sciences              1           4    Male         3  ...   
4          1        Medical              1           5    Male         1  ...   

  Over18 PercentSalaryHike  StandardHours  StockOptionLevel TotalWorkingYears  \
0      Y                11              8                 0               1.0   
1      Y                23              8                 1               6.0   
2      Y                15              8                 3               5.0   
3      Y                11              8                 3              13.0   
4      Y                12              8                 2               9.0   

   TrainingTimesLastYear  YearsAtCompany  YearsSinceLastPromotion  \
0                      6               1                        0   
1                      3               5                        1   
2                      2               5                        0   
3                      5               8                        7   
4                      2               6                        0   

   YearsWithCurrManager  Attrition  
0                     0          0  
1                     4          1  
2                     3          0  
3                     5          0  
4                     4          0  

[5 rows x 24 columns]

# Wilcoxon - Sign Test
### Distance from home and StandardHours

In [18]:
from scipy.stats import wilcoxon

stats,p=wilcoxon(dataset2.DistanceFromHome,dataset2.StandardHours)
print(stats,p)

4191909.0 0.043962321781016465


### H0: There is no significant differences in the distance from home and StandardHours
### Ha: There is significant differences in the distance from home and StandardHours
### The p value is 0.043962321781016465 ,So p value is less than 0.5
### Hence Ho (null hypothesis) is rejected and Ha (alternative hypothesis) accepted.
### There is significant differences in the distance from home and StandardHours

# Friedman Test
### YearsAtCompany	and YearsSinceLastPromotion	and YearsWithCurrManage

In [19]:
from scipy.stats import friedmanchisquare

stats,p = friedmanchisquare(dataset2.YearsAtCompany,dataset2.YearsSinceLastPromotion,dataset2.YearsWithCurrManager)
print(stats,p)

6499.282582216817 0.0


### H0: There is no significant differences in the YearsAtCompany and YearsSinceLastPromotion and YearsWithCurrManage
### Ha: There is significant differences in the YearsAtCompany and YearsSinceLastPromotion and YearsWithCurrManage
### The p value is 0.0 ,So p value is less than 0.5
### Hence Ho (null hypothesis) is rejected and Ha (alternative hypothesis) accepted.
### There is significant differences in the YearsAtCompany and YearsSinceLastPromotion and YearsWithCurrManage

# Mann-Whiteney Test
### Attrition and  DistanceFromHome

In [21]:
from scipy.stats import mannwhitneyu
stats,p=mannwhitneyu(dataset2.Attrition,dataset2.DistanceFromHome)
print(stats,p)

221832.0 0.0


### H0: There is no significant differences in the attrition yes and no for distance from home
### Ha: There is significant differences in the attrition yes and no for distance from home
### The p value is 0.00 ,So p value is less than 0.5
### Hence Ho (null hypothesis) is rejected and Ha (alternative hypothesis) accepted.
### There is significant differences in the attrition yes and no for distance from home

# Kruskal-Wallis Test
### PercentSalaryHike	and StandardHours	and StockOptionLevel

In [23]:
from scipy.stats import kruskal

stats,p = kruskal(dataset2.PercentSalaryHike,dataset2.StandardHours,dataset2.StockOptionLevel)
print(stats,p)

12286.581699858103 0.0


### H0: There is no significant differences in the PercentSalaryHike and StandardHours and StockOptionLevel
### Ha: There is significant differences in the PercentSalaryHike and StandardHours and StockOptionLevel
### The p value is 0.00 ,So p value is less than 0.5
### Hence Ho (null hypothesis) is rejected and Ha (alternative hypothesis) accepted.
### There is significant differences in the PercentSalaryHike and StandardHours and StockOptionLevel

# Chai-Square Test
### Attrition and Gender

In [25]:
from scipy.stats import chi2_contingency

chitable = pd.crosstab(dataset2.Attrition,dataset2.Gender)
chitable

Gender     Female  Male
Attrition              
0            1494  2205
1             270   441

In [26]:
stats,p,dof,expected=chi2_contingency(chitable)
print(stats,p)

1.349904410246582 0.24529482862926827


### H0: There is no significant differences in the attrition yes and no for Gender
### Ha: There is significant differences in the attrition yes and no for Gender
### The p value is 0.24529482862926827 ,So p value is greater than 0.5
### Hence Ho (null hypothesis) is accepted and Ha (alternative hypothesis) rejected.
### There is no significant differences in the attrition yes and no for Gender.